sale_coupon_multi_use/models/sale_order.py

```diff
@@ -1,32 +1,134 @@
-from odoo import models
+from odoo import fields, models
+
+ORDER_CTX_KEY = "coupon_sale_order"
+
+
+def _get_order_line_coupons_filter(line):
+    return lambda r: r.program_id.discount_line_product_id == line.product_id
 
 
 class SaleOrder(models.Model):
     """Extend to modify action_confirm for multi-use coupons."""
 
     _inherit = "sale.order"
 
-    # TODO: implement multi-use coupon relation on sale.order ->
-    # applied_coupon_ids is o2m, which removes coupon if same coupon
-    # is applied on another SO. But for multi-use coupon, it must be
-    # able to keep same coupon on multiple sale orders.
-    # TODO: action_draft seems buggy, because it does not reset back
-    # coupons to new state (you can abuse coupon this way and use normal
-    # coupon multiple times).
+    coupon_multi_use_ids = fields.Many2many(
+        "sale.coupon",
+        "sale_order_coupon_multi_rel",
+        "sale_id",
+        "coupon_id",
+        string="Multi Use Coupons",
+        copy=False,
+        readonly=True,
+    )
+
+    def _get_multi_use_coupons_to_handle(self, filter_func=None):
+        self.ensure_one()
+        coupons = self.coupon_multi_use_ids - self.applied_coupon_ids
+        if filter_func:
+            return filter_func(coupons)
+        return coupons
```

Comment:
Remove this useless function used only once time and with a parameter never used.

Reply:
Yes, that param is leftover from previous iteration. It can be removed. Though function usefulness should not be defined how many times its called. Its other purpose is SoC. Though I guess in this case, we can remove it.